### PR TITLE
Add support for sparse working copies (#52)

### PR DESCRIPTION
**Pull request by martinvonz**
# Checklist

- [x] I have made relevant updates to `CHANGELOG.md`

### CODE REVIEW

**Comment by martinvonz** [comment]
Note that this depends on #236, so you probably want to review that first

File: lib/src/matchers.rs [hunk 1]
```diff
@@ -19,18 +19,14 @@ use std::collections::{BTreeSet, HashMap, HashSet};
...
-        Self {
-            dirs: VisitDirs::All,
-            files: VisitFiles::All,
-        }
-    }
+pub enum Visit {
+    /// Everything in the directory is *guaranteed* to match, no need to check
+    /// descendants
+    AllRecursively,
+    Some {
```
Comment (spectral54) [inline lib/src/matchers.rs:28]: Is 'Some' a good name, given its normal connotations?

File: lib/src/working_copy.rs [hunk 2]
```diff
@@ -132,6 +134,20 @@ fn file_states_from_proto(
     file_states
 }
 
+fn sparse_patterns_from_proto(proto: &crate::protos::working_copy::TreeState) -> Vec<RepoPath> {
+    let mut sparse_patterns = vec![];
+    if proto.has_sparse_patterns() {
```
Comment (spectral54) [inline lib/src/working_copy.rs:139]: protobufs, if that's what these are, have a difference between "has foo" and "foo is non-empty". Should this be:

```
if proto.has_sparse_patterns() {
  ...
}

// For compatibility with old working copies, and empty sparse patterns.
if sparse_patterns.is_empty() {
  sparse_patterns.push(RepoPath::root());
}

sparse_patterns
```

File: lib/src/working_copy.rs [hunk 3]
```diff
@@ -894,6 +935,23 @@ impl LockedWorkingCopy<'_> {
         self.wc.tree_state().as_mut().unwrap().reset(new_tree)
     }
 
+    pub fn sparse_patterns(&self) -> Vec<RepoPath> {
+        self.wc
```
Comment (spectral54) [inline lib/src/working_copy.rs:980]: `self.wc.sparse_patterns()`?

File: lib/src/working_copy.rs [hunk 4]
```diff
@@ -336,6 +337,10 @@ impl TreeState {
         let mut deleted_files: HashSet<_> = self.file_states.keys().cloned().collect();
         while !work.is_empty() {
             let (dir, disk_dir, git_ignore) = work.pop().unwrap();
+            let visit = sparse_matcher.visit(&dir);
```
Comment (spectral54) [inline lib/src/working_copy.rs:340]: Variable only used on next line, right? Inline in `if`?

File: lib/src/matchers.rs (hunk 1 above)
Comment (martinvonz) [reply lib/src/matchers.rs]: Yes, makes sense to rename it. Went with "Specific".

File: lib/src/working_copy.rs (hunk 2 above)
Comment (martinvonz) [reply lib/src/working_copy.rs:139]: No, the idea is that the presence of the `sparse_pattern` field indicates that the working copy was written by a version after this commit. The goal is to allow and empty list of patterns.

File: lib/src/working_copy.rs (hunk 4 above)
Comment (martinvonz) [reply lib/src/working_copy.rs]: Done

File: lib/src/working_copy.rs (hunk 3 above)
Comment (martinvonz) [reply lib/src/working_copy.rs]: ah, much better :)

**Review by spectral54 — APPROVED** [review]

**Pull request merged**